modules/core/src/views/view-manager.js

```diff
@@ -232,26 +229,42 @@ export default class ViewManager {
     if (updateReason) {
       const {width, height, views} = this;
 
+      const oldControllers = this.controllers;
+      this.controllers = {};
+
       this._viewports = views.map(view => {
         const viewState = this.getViewState(view.id);
         const viewport = view.makeViewport({width, height, viewState});
 
-        const controller = this.controllers[view.id];
-        // TODO - check for removed view ids?
-        if (controller) {
-          controller.setProps(
-            Object.assign({}, view.controller, viewState, {
-              x: viewport.x,
-              y: viewport.y,
-              width: viewport.width,
-              height: viewport.height
-            })
-          );
+        // Update the controller
+        if (view.controller) {
+          const controllerProps = Object.assign({}, view.controller, view.defaultState, viewState, {
+            id: view.id,
+            x: viewport.x,
+            y: viewport.y,
+            width: viewport.width,
+            height: viewport.height
+          });
+
+          let controller = oldControllers[view.id];
+          if (controller) {
+            controller.setProps(controllerProps);
+          } else {
+            controller = this._createController(controllerProps);
+          }
+          this.controllers[view.id] = controller;
         }
 
         return viewport;
       });
 
+      // Remove unused controllers
+      for (const id in oldControllers) {
+        if (!this.controllers[id]) {
+          oldControllers[id].finalize();
+        }
+      }
+
```

Comment:
nit: add `oldControllers = {}` to remove holding any object references.

Reply:
`oldControllers` should be GC'ed at the end of this function.